### PR TITLE
RFC: Add in contract transfer support 

### PR DESCRIPTION
**Pull request by chenfengjin**

### CODE REVIEW

File: kernel/contract/state.go [hunk 1]
```diff
@@ -31,6 +38,8 @@ type XMState interface {
 
 // XMState 对XuperBridge暴露对账本的UTXO操作能力
 type UTXOState interface {
+	Transfer(from string, to string, amount *big.Int) error
+	GetUtxoRWSets() ([]*protos.TxInput, []*protos.TxOutput)
```
Comment (icexin) [inline kernel/contract/state.go:42]: 放到StateSandbox里面

File: kernel/contract/sandbox/xmcache.go [hunk 2]
```diff
@@ -45,18 +48,19 @@ type XMCache struct {
...
 
-	// utxoCache       *UtxoCache
+	utxoSandbox *utxo.UTXOSandbox
 	// crossQueryCache *CrossQueryCache
 	events []*protos.ContractEvent
 }
 
 // NewXModelCache new an instance of XModel Cache
-func NewXModelCache(model ledger.XMReader) *XMCache {
+func NewXModelCache(model ledger.XMReader, vm UtxoVM, utxoInput []*protos.TxInput, Penetrate bool) *XMCache {
```
Comment (icexin) [inline kernel/contract/sandbox/xmcache.go:57]: XMdelCache和utxoSandbox分开，放到一个StateSandbox里面

File: kernel/engines/xuperos/chain.go [hunk 3]
```diff
@@ -130,7 +130,9 @@ func (t *Chain) PreExec(ctx xctx.XContext, reqs []*protos.InvokeRequest, initiat
 	}
 
 	stateConfig := &contract.SandboxConfig{
-		XMReader: t.ctx.State.CreateXMReader(),
+		XMReader:  t.ctx.State.CreateXMReader(),
+		UtxoVM:    t.ctx.State,
```
Comment (icexin) [inline kernel/engines/xuperos/chain.go:134]: UtxoVM改为UtxoReader

File: kernel/contract/state.go (hunk 1 above)
Comment (chenfengjin) [reply kernel/contract/state.go]: Done

File: kernel/engines/xuperos/chain.go (hunk 3 above)
Comment (chenfengjin) [reply kernel/engines/xuperos/chain.go]: Done

File: kernel/contract/sandbox/xmcache.go (hunk 2 above)
Comment (chenfengjin) [reply kernel/contract/sandbox/xmcache.go]: Done

File: bcs/ledger/xledger/state/tx_verification.go [hunk 4]
```diff
@@ -565,8 +565,14 @@ func (t *State) verifyTxRWSets(tx *pb.Transaction) (bool, error) {
 	}
 
 	reader := sandbox.XMReaderFromRWSet(rwSet)
+	utxoInput, err := xmodel.ParseContractUtxoInputs(tx)
+	if err != nil {
+		return false, err
+	}
+	utxoReader := utxo.NewUTXOReaderFromInput(utxoInput)
```
Comment (icexin) [inline bcs/ledger/xledger/state/tx_verification.go:572]: NewUTXOReaderFromInput放到sandbox里面会好些

**Review by icexin — APPROVED** [review]

**Review by KenianShi — APPROVED** [review]

**Review by qizheng09 — APPROVED** [review]

**Pull request merged**